image/image_linux.go

```diff
@@ -473,12 +474,50 @@ func (s *imageSeeder) deriveInfoForLocalSnaps(f seedwriter.SeedAssertionFetcher,
 			return nil, err
 		}
 
+		// Search matching local components
+		seedComps := []seedwriter.SeedComponent{}
+		for path, compInfo := range pathToLocalComp {
+			if compInfo.Component.SnapName != info.SnapName() {
+				continue
+			}
+
+			// Check if the component is defined by the snap
+			compInSnap, ok := info.Components[compInfo.Component.ComponentName]
+			if !ok {
+				return nil, fmt.Errorf("component %s is not defined by snap %s",
+					compInfo.Component.ComponentName, info.SnapName())
+			}
+			// and if types match
+			if compInSnap.Type != compInfo.Type {
+				return nil, fmt.Errorf("component %s has type %s while snap %s defines type %s for it",
+					compInfo.Component.ComponentName, compInfo.Type, info.SnapName(), compInSnap.Type)
+			}
+			// Add component now to SeedSnap
+			seedComps = append(seedComps, seedwriter.SeedComponent{
+				ComponentRef: compInfo.Component,
+				Path:         path,
+				Info:         compInfo,
+			})
+			delete(pathToLocalComp, path)
+		}
+
+		s.w.AddComponentsToSnap(sn, seedComps)
```

Comment:
I'm also a bit unsure about this API, shouldn't some of these checks be done by the writer instead?

Reply:
Agreed, I've moved some checks to the writer.